feat(Probability/Distributions): formalize Beta distribution

arulandu opened the pull request:
Formalize Beta distribution, using Gamma distribution as a reference. Added real-valued beta wrapper, in the manner of gamma. Thanks to @EtienneC30 for help with casting real <-> complex.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> Mathlib.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

arulandu committed:
beta scaffold

arulandu committed:
purge sorrys

arulandu committed:
no gamma zero dependency

arulandu committed:
scaffold

arulandu committed:
casts + bochner left

arulandu committed:
one casting proof

arulandu committed:
need open interval cast / integrability cast proofs

arulandu committed:
formalized! thanks entienne for casting

arulandu commented on Mathlib/Probability/Distributions/Beta.lean:
```diff
@@ -0,0 +1,331 @@
+  exact h
+
+lemma betaIntervalIntegralPos {a b : ℝ} (ha: 0 < a) (hb : 0 < b) :
+∫ (x : ℝ) in (0)..1, x ^ (a - 1) * (1 - x) ^ (b - 1) > 0 := by
```
bot fix style

arulandu committed:
move beta special func defs

arulandu committed:
lake exe mk

Commit:
commit changes from style linters

arulandu committed:
lint

arulandu committed:
simpler beta_integrand_intervalIntegrable

EtienneC30 commented on Mathlib/Analysis/SpecialFunctions/Gamma/Beta.lean:
```diff
@@ -458,10 +458,95 @@ theorem Gamma_ne_zero_of_re_pos {s : ℂ} (hs : 0 < re s) : Gamma s ≠ 0 := by
   contrapose! hs
   simpa only [hs, neg_re, ← ofReal_natCast, ofReal_re, neg_nonpos] using Nat.cast_nonneg _
 
+lemma integral_re {X : Type*} [MeasurableSpace X]
+    (μ : Measure X) {f : X → ℂ} (hf : Integrable f μ) :
+    ∫ x, (f x).re ∂μ = (∫x, f x ∂μ).re := _root_.integral_re hf
+
+lemma setIntegral_re {X : Type*} [MeasurableSpace X]
+    (μ : Measure X) {s : Set X} {f : X → ℂ} (hf : IntegrableOn f s μ) :
+    ∫ x in s, (f x).re ∂μ = (∫x in s, f x ∂μ).re := Complex.integral_re _ hf.integrable
+
```
You should open a separate PR for these, as they have nothing to do with the `Beta` function, and make this PR depend on it (the hidden comment in the PR description explains how to indicate a dependency). I would put them in `Mathlib.MeasureTheory.Integral.SetIntegral`, along their `RCLike` versions. Please also add an `intervalIntegral` version.

EtienneC30 commented on Mathlib/Analysis/SpecialFunctions/Gamma/Beta.lean:
Again you should open a new PR and make this one depend on it. When modifying different files, it is almost always worth to open a separate PR for each file, it makes reviewing much easier.

EtienneC30 commented on Mathlib/Analysis/SpecialFunctions/Gamma/Beta.lean:
```diff
@@ -458,10 +458,95 @@ theorem Gamma_ne_zero_of_re_pos {s : ℂ} (hs : 0 < re s) : Gamma s ≠ 0 := by
+def Beta (a b : ℝ) : ℝ :=
+  (Complex.betaIntegral a b).re
+
+lemma integrable_iff_ofReal {X : Type*} [MeasurableSpace X] {μ : Measure X}
+    {f : X → ℝ} : Integrable f μ ↔ Integrable (fun x ↦ (f x : ℂ)) μ :=
+    ⟨fun hf ↦ hf.ofReal, fun hf ↦ hf.re⟩
+
+lemma integrableOn_iff_ofReal {X : Type*} [MeasurableSpace X] {μ : Measure X} {s : Set X}
+    {f : X → ℝ} : IntegrableOn f s μ ↔ IntegrableOn (fun x ↦ (f x : ℂ)) s μ := integrable_iff_ofReal
```
Please open a fourth PR for these, again it is independent from the rest. I'd put the first in `Mathlib.MeasureTheory.Function.L1Space`, next to `Integrable.ofReal`, the second in `Mathlib.MeasureTheory.Integral.IntegrableOn`, and add an `IntervalIntegrable` version please in `Mathlib.MeasureTheory.Integral.IntervalIntegral`.

EtienneC30 replied:
In fact you can put these results and the ones about `integral_re` in the same PR.

EtienneC30 commented on Mathlib/Analysis/SpecialFunctions/Gamma/Beta.lean:
```diff
@@ -458,10 +458,95 @@ theorem Gamma_ne_zero_of_re_pos {s : ℂ} (hs : 0 < re s) : Gamma s ≠ 0 := by
+
+lemma BetaIntegral_ofReal {a b : ℝ} (ha : 0 < a) (hb : 0 < b) :
+    Beta a b = ∫ (x : ℝ) in Ioo 0 1, x ^ (a - 1) * (1 - x) ^ (b - 1) := by
+  rw [Beta, Complex.betaIntegral]
+  rw [BetaIntegral_ofReal_interval ha hb]
+  rw [intervalIntegral.integral_of_le (by norm_num)]
+  rw [integral_Ioc_eq_integral_Ioo]
```
You can put all these in one `rw`:
```suggestion
rw [Beta, Complex.betaIntegral, BetaIntegral_ofReal_interval ha hb,
    intervalIntegral.integral_of_le (by norm_num), integral_Ioc_eq_integral_Ioo]
```

EtienneC30 commented on Mathlib/Analysis/SpecialFunctions/Gamma/Beta.lean:
```diff
@@ -458,10 +458,95 @@ theorem Gamma_ne_zero_of_re_pos {s : ℂ} (hs : 0 < re s) : Gamma s ≠ 0 := by
+  rw [intervalIntegral.integral_of_le (by norm_num)]
+  rw [integral_Ioc_eq_integral_Ioo]
+
+private lemma Beta_integrable_cast {a b : ℝ} :
+    IntervalIntegrable (fun x ↦ (x : ℂ) ^ (↑a - 1:ℂ) * (1 - ↑x) ^ (↑b - 1:ℂ)) volume 0 1 ↔
+    IntervalIntegrable (fun x ↦ x ^ (a - 1) * (1 - x) ^ (b - 1)) volume 0 1 := by
```
Again I think it's better if you get rid of this lemma and integrate it in the proof below.

arulandu replied:
agreed, done

EtienneC30 commented on Mathlib/Analysis/SpecialFunctions/Gamma/Beta.lean:
```diff
@@ -458,10 +458,95 @@ theorem Gamma_ne_zero_of_re_pos {s : ℂ} (hs : 0 < re s) : Gamma s ≠ 0 := by
+  exact h
+
+lemma Beta_pos {a b : ℝ} (ha : 0 < a) (hb : 0 < b) : Beta a b > 0 := by
+  rw [Beta, Complex.betaIntegral]
+  rw [BetaIntegral_ofReal_interval ha hb]
```
Again you can put everything in one `rw`

EtienneC30 commented on Mathlib/Analysis/SpecialFunctions/Gamma/Beta.lean:
```diff
@@ -458,10 +458,95 @@ theorem Gamma_ne_zero_of_re_pos {s : ℂ} (hs : 0 < re s) : Gamma s ≠ 0 := by
+  rw [← intervalIntegrable_iff_integrableOn_Ioc_of_le (by norm_num)]
+  exact Complex.betaIntegral_convergent ha' hb'
+
+lemma BetaIntegral_ofReal {a b : ℝ} (ha : 0 < a) (hb : 0 < b) :
+    Beta a b = ∫ (x : ℝ) in Ioo 0 1, x ^ (a - 1) * (1 - x) ^ (b - 1) := by
```
Why don't you write it as an integral in `(0)..1`?

arulandu replied:
fixed

EtienneC30 reviewed (commented):
That's nice, but some lemmas are unnecessary in my opinion. Also, even though I reviewed this, please open another PR and make the modifications there.

EtienneC30 commented on Mathlib/Probability/Distributions/Beta.lean:
```diff
@@ -0,0 +1,213 @@
+    have h_betapos := Beta_pos ha hb
+    positivity
+  · rw [EventuallyLE, ae_restrict_iff' measurableSet_Ioo]
+    exact ae_of_all _ (fun x (hx : 0 < x ∧ x < 1) ↦ by
+      rw [Pi.zero_apply]
+      have hpos := betaPDFReal_nonneg ha hb
+      unfold betaPDFReal at hpos
+      specialize hpos x
+      rw [if_pos hx] at hpos
+      exact hpos
+    )
```
Here is a bit shorter way, feel free to add it or not. However, please avoid to write multiline tactic blocks inside the arguments of a theorem. Use `apply` or `refine` instead.
```suggestion
    refine ae_of_all _ fun x hx ↦ ?_
    convert betaPDFReal_nonneg ha hb x using 1
    rw [betaPDFReal, if_pos (mem_Ioo.1 hx)]
```

arulandu replied:
ah got it

EtienneC30 commented on Mathlib/Probability/Distributions/Beta.lean:
```diff
@@ -0,0 +1,213 @@
+    )
+  · apply (measurable_betaPDFReal a b).aestronglyMeasurable.congr
+    refine (ae_restrict_iff' measurableSet_Ioo).mpr <| ae_of_all _ fun x (hx : 0 < x ∧ x < 1) ↦ ?_
+    simp_rw [betaPDFReal, eq_true_intro hx, ite_true]
+end BetaPDF
```
```suggestion
    simp_rw [betaPDFReal, eq_true_intro hx, ite_true]

end BetaPDF
```

EtienneC30 reviewed (commented):
That's a good job. My comments are mostly about conventions in naming, writing proofs etc. I sometimes put shorter writing, feel free to use it if you like it or discard it if you think it makes code less readable.

arulandu commented on Mathlib/Analysis/SpecialFunctions/Gamma/Beta.lean:
```diff
@@ -458,10 +458,95 @@ theorem Gamma_ne_zero_of_re_pos {s : ℂ} (hs : 0 < re s) : Gamma s ≠ 0 := by
+lemma integrableOn_iff_ofReal {X : Type*} [MeasurableSpace X] {μ : Measure X} {s : Set X}
+    {f : X → ℝ} : IntegrableOn f s μ ↔ IntegrableOn (fun x ↦ (f x : ℂ)) s μ := integrable_iff_ofReal
+
+lemma BetaIntegral_ofReal_interval {a b : ℝ} (ha : 0 < a) (hb : 0 < b) :
+    (∫ (x : ℝ) in (0)..1, (x : ℂ) ^ (↑a - 1:ℂ) * (1 - ↑x) ^ (↑b - 1:ℂ)).re =
+      ∫ (x : ℝ) in (0)..1, x ^ (a - 1) * (1 - x) ^ (b - 1) := by
```
The reason I had this separated was since it was a fairly large chunk of code repeated btw the following lemma and Beta_pos (which I thought was a nice fact to add in the special functions). I will put this in that PR as opposed to this one

EtienneC30 replied:
But in `Beta_pos` you can directly `rw` with `BetaIntegral_ofReal`.

vihdzp replied:
If you have a result that's important in the context of a proof but artificial or useless elsewhere, usually the solution is to make it a `private` lemma or theorem.

EtienneC30 replied:
I don't think this is the way to go here, as this lemma is the main part of the proof `BetaIntegral_ofReal` and is not used elsewhere (as I said, it is not necessary for `Beta_pos`), so I think it definitely makes more sense to put it directly in the proof of `BetaIntegral_ofReal`.

arulandu commented on Mathlib/Probability/Distributions/Beta.lean:
```diff
@@ -0,0 +1,213 @@
+    measurable_const
+
+/-- The beta pdf is strongly measurable -/
+@[measurability]
+ lemma stronglyMeasurable_betaPDFReal (a b : ℝ) :
+     StronglyMeasurable (betaPDFReal a b) :=
+   (measurable_betaPDFReal a b).stronglyMeasurable
```
unrecognized function property (same as for pareto PR)

EtienneC30 replied:
And same answer, sorry for those

arulandu commented on Mathlib/Probability/Distributions/Beta.lean:
```diff
@@ -0,0 +1,213 @@
+
+section BetaPDF
+
+/-- The pdf of the beta distribution depending on its scale and rate -/
```
got it, apologies

arulandu commented on Mathlib/Probability/Distributions/Beta.lean:
```diff
@@ -0,0 +1,213 @@
+    have h2 : (1 - x) ^ (b - 1) ≥ 0 := by
+      have hp : 1 - x ≥ 0 := by linarith
+      exact rpow_nonneg hp (b - 1)
+    have hbeta:= Beta_pos ha hb
+
+    positivity
```
fixed all of these

arulandu commented on Mathlib/Probability/Distributions/Beta.lean:
```diff
@@ -0,0 +1,213 @@
+
+open MeasureTheory Real Set Filter Topology
+
+/-- A Lebesgue Integral from -∞ to y can be expressed as the sum of one from -∞ to 0 and 0 to x -/
+lemma lintegral_Iic_eq_lintegral_Iio_add_Icc {y z : ℝ} (f : ℝ → ℝ≥0∞) (hzy : z ≤ y) :
+    ∫⁻ x in Iic y, f x = (∫⁻ x in Iio z, f x) + ∫⁻ x in Icc z y, f x := by
+  rw [← Iio_union_Icc_eq_Iic hzy, lintegral_union measurableSet_Icc]
+  rw [Set.disjoint_iff]
+  rintro x ⟨h1 : x < _, h2, _⟩
+  linarith
```
Yeah, I don't actually use this so it's clutter. If I were to separate PR this + the other lintegral_eq lemma that I do use, where would recommend I put this?

arulandu commented on Mathlib/Probability/Distributions/Beta.lean:
```diff
@@ -0,0 +1,213 @@
+open Measure
+
+/-- A Lebesgue Integral from -∞ to y can be expressed as the sum of one from -∞ to 0 and 0 to x -/
+lemma lintegral_Iio_eq_lintegral_Iic_add_Ioo {y z : ℝ} (f : ℝ → ℝ≥0∞) (hzy : z < y) :
+    ∫⁻ x in Iio y, f x = (∫⁻ x in Iic z, f x) + ∫⁻ x in Ioo z y, f x := by
```
got it

arulandu committed:
review: proof fixes

arulandu committed:
typos

arulandu commented:
@EtienneC30 I've addressed your comments about writing etc. For the new PRs, I just wanted to condense your feedback and make sure I understand this correctly:
PR1: `integrable_iff_ofReal` and similar for integrableOn, each in the file locations you suggested. integral_re and setIntegral_re as well. make equivalents for intervalIntegrable
PR2: `def Beta`, Beta_pos, BetaIntegral_ofReal, Beta_integrand_intervalIntegrable in `SpecialFunctions/Gamma/Beta.lean`
PR3: Beta formalization in `/Distributions/Beta.lean` 

Dependencies: (3) <- (1) + (2) and (2) <- (1). This PR will then get dropped in favor for 3, after 1 and 2 are merged. 

I understand having the dependencies specified in the PR, but in order for the code in the branch respective to PR3 for example to run, won't the PR1 + PR2 changes need to be present, meaning the diff will have multiple files? Or should I just do PR1, wait for finalization then PR2, then PR3?

arulandu committed:
fix proof error

arulandu committed:
lint

arulandu committed:
interval integral

EtienneC30 commented:
You understood perfectly (indeed it was not very clear sorry for that). The way to go is as follows: indeed your modifications of PR1 and PR2 will appear in PR3 as well, making more diffs. But it does not matter because PR3 won't be reviewed while PR1 and PR2 aren't. When for example PR1 is merged to master, you should merge master into the branch of your PR2, that way the old diff of PR1 disappears. And once PR2 is merged to master, you can merge master to the branch of PR3.

arulandu commented:
Ok if this is the case, it seems easiest to just do these PR's one at a time, since there may be changes in the review of PR 1 etc. Will close this PR and open 1, then 2, then 3
